### DISCUSSION TITLE
Help needed on "CS8774 Member must have a non-null value when exiting"

### DISCUSSION THREAD

**Discussion by qrli** [dotnet/csharplang#4270] [Q&A] [answered]
Need some help on `[MemberNotNull]`. I'm using Dapper.
```cs
        [MemberNotNull(nameof(dqListCache))]
        public async Task InitCacheAsync()
        {
            using var conn = connFactory.NewConnection();
            dqListCache = (await conn.QueryAsync<MyDto>("select * from MyTable")).ToList();
        }
```
It generates "CS8774 Member must have a non-null value when exiting." on the last line, and adding null-forgiving `!` anywhere cannot mute it.

**Comment by CyrusNajmabadi** [comment]
It's possible that this is because it's task returning and async.  So the method literally can return (giving the caller the task) before the value is actually set.

**Comment by CyrusNajmabadi** [comment, marked as answer]
What happens if you start the method with `dqListCache = null!` prior to any of the existing code.

**Reply by qrli** [reply]
Cool. It resolves the warning.

A bit counter intuitive, but it does work as defined. I expected `[MemberNotNull]` to remove null-forgiving from my code, but it seems it needs it by itself. 😂 `async` initialization seems a bit tricky here and there.
